query/common/src/main/java/org/codice/ddf/admin/common/PrioritizedBatchExecutor.java

```diff
@@ -0,0 +1,228 @@
+                    "Argument {threadPoolSize} with value [{}] exceeds maximum allowed value, defaulting to the max number of threads [{}].",
+                    threadPoolSize,
+                    MAX_THREAD_POOL_SIZE);
+
+            threadPoolSize = MAX_THREAD_POOL_SIZE;
+        }
+
+        this.tasks = tasks;
+        this.taskHandler = taskHandler;
+
+        threadPool = Executors.newFixedThreadPool(threadPoolSize);
+    }
+
+    /**
+     * @return an {@code Optional} containing a task's result, if there was one
+     * @see #getFirst(long, TimeUnit)
+     */
+    public Optional<R> getFirst() {
+        return getFirst(DEFAULT_WAIT_TIME_SEC, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Start task execution and blocks until the highest priority task batch has returned a valid result
+     * according to the task handler, then cleans up remaining tasks. The current instance of the
+     * {@code PrioritizedBatchExecutor} is not useable after calling {@code getFirst(long, TimeUnit)}.
+     * <p>
+     * If the {@code totalWaitTime} is exceeded, no result has been found yet, and all batches have not been polled,
+     * each remaining batch will be polled at least once until a result is found or until all batches have been
+     * polled.
+     *
+     * @param totalWaitTime total wait time for execution
+     * @param timeUnit      {@code TimeUnit} to use for the {@code batchWaitTime}
+     * @return an {@code Optional} containing a task's result, if there was one
+     */
+    public Optional<R> getFirst(long totalWaitTime, TimeUnit timeUnit) {
+        Validate.isTrue(totalWaitTime >= 1, "Batch wait time must be greater than 0.");
+        Validate.notNull(timeUnit, "Argument {timeUnit} cannot be null.");
+
+        try {
+            List<CompletionService<T>> prioritizedCompletionServices =
+                    getPrioritizedCompletionServices();
+
+            long totalWaitTimeMillis = TimeUnit.MILLISECONDS.convert(totalWaitTime, timeUnit);
+            long endTime = System.currentTimeMillis() + totalWaitTimeMillis;
+
+            for (int i = 0; i < tasks.size(); i++) {
+                LOGGER.debug("Executing batch {}.", i + 1);
+
+                CompletionService<T> completionService = prioritizedCompletionServices.get(i);
+                int currentBatchSize = tasks.get(i)
+                        .size();
+
+                long lastBatchPollTime = System.currentTimeMillis();
+                for (int j = 0; j < currentBatchSize; j++) {
+
+                    Future<T> taskFuture;
+
+                    if (lastBatchPollTime >= endTime) {
+                        LOGGER.debug("\tExceeded max wait time of {} {}. Polling remaining batches.",
+                                totalWaitTime,
+                                timeUnit.toString());
+
+                        while ((taskFuture = completionService.poll()) != null) {
+                            Optional<R> result = handleTaskResult(taskFuture);
```

Comment:
✏️ I would recommend refactoring lines 137-144 into a method which you could then also use for the block of code from 168-175.